File: FindDiagonalOrder.py
from pip import List
#Time Complexity = O(n)
#Space Complexity = O(1)

class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        if(len(mat)==0):
            return []
        m = len(mat)
        n = len(mat[0])
        res = [0] * (m*n)
        r = 0
        c = 0
        index = 0
        dir = 1
        while(index<(m*n)):
            res[index]=mat[r][c]
            if(dir == 1):
                if(c == n-1):
                    dir = -1
                    r+=1
                elif(r==0):
                    dir = -1
                    c+=1
                else:
                    r-=1
                    c+=1
            else:
                if(r == m-1):
                    dir = 1
                    c+=1
                elif(c==0):
                    dir = 1
                    r+=1
                else:
                    c-=1
                    r+=1
            index+=1
        return res
                
            
        
        

File: FindDiagonalOrder.cs
// Time Complexity : O(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode :
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach

public int[] FindDiagonalOrder(int[][] mat) {
        
     if (mat == null || mat.Length == 0)
            return new int[] { };

        int m = mat.Length;
        int n = mat[0].Length;
        int[] result = new int[m * n];

        int r = 0;
        int c = 0;
        int dir = 1; ///1 is upward and -1 is downward
        int i = 0;

        while (r < m && c < n)
        {   
            result[i] = mat[r][c];
            
            if(dir == 1) //going up
            {
                if(c == n-1)
                {
                    dir = -1;
                    r++;
                }
                else if(r == 0)
                {
                    dir = -1;
                    c++;
                }
                else
                {
                    r--;
                    c++;
                }
            }
            else //going down
            {
                if(r == m-1)
                {
                    dir = 1;
                    c++;
                }
                else if(c == 0)
                {
                    dir = 1;
                    r++;
                }
                else
                {
                    r++;
                    c--;
                }
            }
            
            i++;
        }
        
        return result;
    }